_posts/2021-11-18-keras-updates/keras-updates.Rmd

```diff
@@ -0,0 +1,159 @@
+
+[^4]: We're not implying they don't exist anymore; they do. But the preferred method is now `%py_class%`.
+
+The new way, `%py_class%`, now allows for translating the above code like this:
+
+```{r}
+NonNegative(keras$constraints$Constraint) %py_class% {
+  "__call__" <- function(x) {
+    w * k_cast(w >= 0, k_floatx())
+  }
+}
+```
+
+Using `%py_class%`, we directly subclass the *Python* `object tf.keras.constraints.Constraint`, and override its `__call__` method.
+
+Why is this so powerful? The first advantage is clearly visible from the example: Translating Python code becomes a mechanical, almost, task. But there's more: The above method is independent from what *kind* of object you're subclassing. Want to implement a new layer? A callback? A loss? An optimizer? The procedure is always the same. No need to find a pre-defined R6 object in the `keras` codebase; one `%py_class%` delivers them all.
+
+There is a lot more to say on this topic, though; in fact, if you don't *want* to use directly, there are wrappers available for the most frequent use cases. More on this in a dedicated post. Until then, consult the [vignette](https://github.com/rstudio/keras/blob/master/vignettes/new-guides/python_subclasses.Rmd) for numerous examples, syntactic sugar, and low-level details.
+
+### RNN cell API
+
+Our third point is at least half as much shout-out to excellent documentation as alert to a new feature. The piece of documentation in question is a new [vignette on RNNs](https://github.com/rstudio/keras/blob/master/vignettes/new-guides/working_with_rnns.Rmd). The vignette gives a useful overview of how RNNs function in Keras, addressing the usual questions that tend to come up once you haven't been using them in a while: What exactly are states vs. outputs, and when does a layer return what? How do I initialize the state in an application-dependent way? What's the difference between stateful and stateless RNNs?
+
+In addition, the vignette covers more advanced questions: How do I pass nested data to an RNN? How do I write custom cells?
+
+In fact, this latter question brings us to the new feature we wanted to call out: the new cell-level API. Conceptually, with RNNs, there's always two things involved: The logic of what happens at a single timestep; and the threading of state across timesteps. So-called "simple RNNs" are concerned with the latter (recursion) aspect only; they tend to exhibit the classic vanishing-gradients problem. Gated architectures, such as the LSTM and the GRU, have specially been designed to avoid those problems; both can be easily integrated into a model using the respective `layer_x()` constructors. What if you'd like, not a GRU, but *something like* a GRU (using some fancy new activation method, say)?
+
+Since Keras 2.7, you can now create a single-timestep RNN cell (using the above-described `%py_class%` API), and obtain a recursive version -- a complete layer -- using `layer_rnn()`:
```

Comment:
```suggestion
With Keras 2.7, you can now create a single-timestep RNN cell (using the above-described `%py_class%` API), and obtain a recursive version -- a complete layer -- using `layer_rnn()`:
```

Reply:
"Since" is one of those words that requires certain verb tenses afterwards. Basically, I think this is a simpler way to go. I think a version usually implies forwards compatibility, but, if not, you could do "Starting with Keras 2.7"

Reply:
hm, yeah ... it did feel wrong in fact, somehow ...